lib/log_displayer.js

```diff
@@ -6,61 +6,73 @@ let url = require('url')
+    if (u.query && u.query.includes('srv')) {
+      return readLogsV1(logplexURL)
+    } else {
+      return readLogsV2(logplexURL)
+    }
   }
-}
 
-function readLogsV1 (logplexURL) {
-  return new Promise(function (resolve, reject) {
-    let res = cli.got.stream(logplexURL)
-    res.setEncoding('utf8')
-    liner.setEncoding('utf8')
-    res.pipe(liner)
-    liner.on('data', line => cli.log(colorize(line)))
-    res.on('end', resolve)
-    res.on('error', reject)
-  })
-}
-
-function readLogsV2 (logplexURL) {
-  return new Promise(function (resolve, reject) {
-    const u = url.parse(logplexURL, true)
-    const isTail = u.query.tail && u.query.tail === 'true'
-    const userAgent = process.env.HEROKU_DEBUG_USER_AGENT || 'heroku-run'
-    const proxy = process.env.https_proxy || process.env.HTTPS_PROXY
-    const es = new EventSource(logplexURL, {
-      proxy,
-      headers: {
-        'User-Agent': userAgent
-      }
+  function readLogsV1 (logplexURL) {
+    return new Promise(function (resolve, reject) {
+      let res = cli.got.stream(logplexURL)
+      res.setEncoding('utf8')
+      liner.setEncoding('utf8')
+      res.pipe(liner)
+      liner.on('data', line => cli.log(colorize(line)))
+      res.on('end', resolve)
+      res.on('error', reject)
     })
+  }
 
-    es.onerror = function (err) {
-      if (!isTail) {
-        resolve()
-        es.close()
-      }
+  function readLogsV2 (logplexURL) {
+    return new Promise(function (resolve, reject) {
+      const u = url.parse(logplexURL, true)
+      const isTail = u.query.tail && u.query.tail === 'true'
+      const userAgent = process.env.HEROKU_DEBUG_USER_AGENT || 'heroku-run'
+      const proxy = process.env.https_proxy || process.env.HTTPS_PROXY
+      const es = new EventSource(logplexURL, {
+        proxy,
+        headers: {
+          'User-Agent': userAgent
+        }
+      })
+      let lines = 0
 
-      if (err && (err.status === 404 || err.status === 403)) {
-        reject(new Error('Log stream timed out. Please try again.'))
-        es.close()
+      es.onerror = function (err) {
+        if (!isTail) {
+          if (lines === 0) {
+            // if app emits too many logs logplex will not return historical logs
+            // in this case, tail instead
+            logDisplayer(heroku, {
+              ...options,
+              tail: true
+            }).then(resolve).catch(reject)
+            return
+          }
+          resolve()
+          es.close()
+        }
+
+        if (err && (err.status === 404 || err.status === 403)) {
+          reject(new Error('Log stream timed out. Please try again.'))
+          es.close()
+        }
       }
-    }
 
-    es.onmessage = function (e) {
-      e.data.trim().split(/\n+/).forEach((line) => {
-        cli.log(colorize(line))
-      })
-    }
-  })
-}
+      es.onmessage = function (e) {
+        e.data.trim().split(/\n+/).forEach(line => {
+          cli.log(colorize(line))
+          lines++
```

Comment:
Might be better to use a boolean here, since I think this is only ever used as a `lines === 0` check. I'm imagining this int will get _very_ large over time.

Reply:
no we use it down below too. And it doesn't matter if it's an int or a boolean, they're both actually integers and it doesn't matter how large it is it takes the same space. Technically it will overflow at 9,223,372,036,854,775,807 but I don't think we need to worry about that.